### ISSUE TITLE
Meter version in Getting Started = Library version - does this makes sense?

### ISSUE THREAD

**Issue by asafm** [open-telemetry/opentelemetry.io#3941]
**What needs to be changed?** 
The Getting Started section says:

> In order to create traces or metrics, you’ll need to first create a tracer and/or meter provider. In general, we recommend that the SDK should provide a single default provider for these objects. You’ll then get a tracer or meter instance from that provider, and give it a name and version. The name you choose here should identify what exactly is being instrumented – if you’re writing a library, for example, then you should name it after your library (for example com.legitimatebusiness.myLibrary) as this name will namespace all spans or metric events produced. It is also recommended that you supply a version string (i.e., semver:1.0.0) that corresponds to the current version of your library or service.

I was under the impression that you shouldn’t place the version of the library or the version of your app in the meter version name.
I thought it signified the version of the metrics in some way.

What is correct?

**What is the name + path of the page that needs changed?** 
https://github.com/open-telemetry/opentelemetry.io/blob/main/content/en/docs/concepts/instrumentation/manual.md#configure-the-opentelemetry-api

**Additional context**:
Related Slack threads with @trask https://cloud-native.slack.com/archives/C014L2KCTE3/p1706822636127569?thread_ts=1706796820.799869&cid=C014L2KCTE3

**Comment by svrnm** [comment]
Thanks for raising this issue @asafm 

We follow the suggestions from the specification here, see the following:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md?plain=1#L133

> `version`: Specifies the version of the instrumentation scope if the scope
  has a version (e.g. a library version). Example value: `1.0.0`.

Also, the glossary states the following (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md?plain=1#L155):

> A logical unit of the application code with which the emitted telemetry can be
associated. It is typically the developer's choice to decide what denotes a
reasonable instrumentation scope. The most common approach is to use the
[instrumentation library](#instrumentation-library) as the scope, however other
scopes are also common, e.g. a module, a package, or a class can be chosen as
the instrumentation scope.

> If the unit of code has a version then the instrumentation scope is defined by
the (name,version) pair otherwise the version is omitted and only the name is
used. The name or (name,version) pair uniquely identify the logical unit of the
code that emits the telemetry. A typical approach to ensure uniqueness is to use
fully qualified name of the emitting code (e.g. fully qualified library name or
fully qualified class name).

This is fairly broad, but to me this sounds that the version & name of the library is a possible choice.

**Comment by cartermp** [comment]
Yeah, my personal take here is that you probably don't actually need a version in code:

- It really ought to be a best practices that your telemetry _is a part of your application_, which means that as your app changes versions, so does your telemetry, and these two things are one and the same. A resource with service version is probably the better way to track this
- If you need to version how you instrument separately from your application you're in confuse-o-tron complexity happy fun time land, sorry to say
- Some languages don't have a good way to pull in the version of what you're doing into code in a way that lets you bump a version once

**Comment by cartermp** [comment]
also I may have written some of the docs that say you should add a version, or at least supplied the code samples that imply it, so mea culpa

**Comment by asafm** [comment]
@cartermp Was the initial motivation to help expose the version of the app or perhaps the used-libraries version, when browsing the metrics?

**Comment by svrnm** [comment]
I think things are different for an instrumentation scope within an app (or a library) and for an instrumentation scope that marks the entry into an (external) library/into the instrumentation library.

Imagine the following scenario: I have a service of version 2.3 deployed 10 times. Now for whatever reason 2 of those instances use a depending library "mylibo" in version 1.3, although the other 8 ones use version 1.4, which has an important fix, or creates telemetry differently, then that version becomes very helpful.

**Comment by asafm** [comment]
Maybe we should clarify - version should be used for libraries mostly?

**Comment by svrnm** [comment]
> Maybe we should clarify - version should be used for libraries mostly?

Yes, or more broadly "with things that have a version, like libraries and applications"

**Comment by asafm** [comment]
So I'm a bit confused. If I'm using OTel to instrument an application, should I use the service or meter versions?

For libraries, it seems clear, at least to me, that the library version should be used.

**Comment by cartermp** [comment]
@asafm I'd still recommend that for your own application instrumentation (i.e., instantiating your own Meter or Tracer for your app internals only), then I wouldn't include a version. 

However, any other time - including libraries that are only internal to your application - then a version makes sense.

That's just my recommendation/opinion, though. I think that for an application, code and instrumentation changes all the time without tying to a specific version, and so trying to include one would just make things harder in the long run.

**Comment by asafm** [comment]
I'm trying to figure out what is agreed upon, so we can update the getting started page text.

**Comment by trask** [comment]
sorry all, I misread the original question and thought it was about encoding a version into the instrumentation scope name, e.g. `mylib:1.0.0`.

my only recommendation for the docs would be to change the example version from `semver:1.0.0` to `1.0.0` (I think my brain told me that `semver` was a library name 😭).

> Was the initial motivation to help expose the version of the app or perhaps the used-libraries version, when browsing the metrics?

the motivation is so that you can identify changes in the telemetry caused by changes to the [instrumentation library](https://opentelemetry.io/docs/concepts/glossary/#instrumentation-library)

> If I'm using OTel to instrument an application, should I use the service or meter versions?

I'm not sure what is the meter version?

**Comment by asafm** [comment]
> I'm not sure what is the meter version?

In Meter version I meant the Instrumentation Scope version (Defined as `version` in getMeter at https://opentelemetry.io/docs/specs/otel/metrics/api/#get-a-meter)

Regarding service version - Meant this https://opentelemetry.io/docs/specs/semconv/resource/#service. The semantic resource.

Should an application (not a library) avoid using the meter version, but place the version using the service version?

**Comment by svrnm** [comment]
The application version and the instrumentation scope version _may_ be the same if the instrumentation scope is the "application". I am not really sure if this is good/bad practice.

**Comment by trask** [comment]
> Should an application (not a library) avoid using the meter version, but place the version using the service version?

I'd personally choose to err on the side of sending both (at least for an OSS application like Apache Pulsar which will have many different telemetry consumers, and so hard to know which version metadata will be useful/used or not).

**Comment by asafm** [comment]
Ok guys, thanks a lot for this discussion. It made me understand we should use meter version which should match the library version or the application version. 
This means the original getting started guide text matches that, hence I can close this issue.
